python/data/preprocess/reader.py

```diff
@@ -34,15 +35,151 @@ def __call__(self, images, **kwargs):
+        depth = np.linalg.norm(points, ord=2, axis=1)
+
+        # get angles of all points
+        scan_x = points[:, 0]
+        scan_y = points[:, 1]
+        scan_z = points[:, 2]
+        yaw = -np.arctan2(scan_y, scan_x)
+        pitch = np.arcsin(scan_z / depth)
+
+        # get projections in image coords
+        proj_x = 0.5 * (yaw / np.pi + 1.0)  # in [0.0, 1.0]
+        proj_y = 1.0 - (
+            pitch + abs(self.lower_inclination)) / self.fov  # in [0.0, 1.0]
+
+        # scale to image size using angular resolution
+        proj_x *= self.proj_W  # in [0.0, W]
+        proj_y *= self.proj_H  # in [0.0, H]
+
+        # round and clamp for use as index
+        proj_x = np.floor(proj_x)
+        proj_x = np.minimum(self.proj_W - 1, proj_x)
+        proj_x = np.maximum(0, proj_x).astype(np.int32)  # in [0,W-1]
+        proj_x_copy = np.copy(
+            proj_x
+        )  # save a copy in original order, for each point, where it is in the range image
+
+        proj_y = np.floor(proj_y)
+        proj_y = np.minimum(self.proj_H - 1, proj_y)
+        proj_y = np.maximum(0, proj_y).astype(np.int32)  # in [0,H-1]
+        proj_y_copy = np.copy(
+            proj_y
+        )  # save a copy in original order, for each point, where it is in the range image
+
+        # unproj_range_copy = np.copy(depth)   # copy of depth in original order
+
+        # order in decreasing depth
+        indices = np.arange(depth.shape[0])
+        order = np.argsort(depth)[::-1]
+        depth = depth[order]
+        indices = indices[order]
+        points = points[order]
+        remission = remissions[order]
+        proj_y = proj_y[order]
+        proj_x = proj_x[order]
+
+        # projected range image - [H,W] range (-1 is no data)
+        proj_range = np.full((self.proj_H, self.proj_W), -1, dtype=np.float32)
+
+        # projected point cloud xyz - [H,W,3] xyz coord (-1 is no data)
+        proj_xyz = np.full((self.proj_H, self.proj_W, 3), -1, dtype=np.float32)
+
+        # projected remission - [H,W] intensity (-1 is no data)
+        proj_remission = np.full((self.proj_H, self.proj_W),
+                                 -1,
+                                 dtype=np.float32)
+
+        # projected index (for each pixel, what I am in the pointcloud)
+        # [H,W] index (-1 is no data)
+        proj_idx = np.full((self.proj_H, self.proj_W), -1, dtype=np.int32)
+
+        proj_range[proj_y, proj_x] = depth
+        proj_xyz[proj_y, proj_x] = points
+        proj_remission[proj_y, proj_x] = remission
+        proj_idx[proj_y, proj_x] = indices
+        proj_mask = proj_idx > 0  # mask containing for each pixel, if it contains a point or not
+
+        data = np.concatenate([
+            proj_range[None, ...],
+            proj_xyz.transpose([2, 0, 1]), proj_remission[None, ...]
+        ])
+        
+        kwargs.update({"samples": data})
+        
+        np.save("proj_mask.npy", proj_mask.astype(np.float32))
```

Comment:
We can pass parameters to the next op, therefore, there is no need to save the file to transfer values.